File: src/event/Event.php
<?php

namespace Ething\Event;


use \Ething\RuleItem;
use \Ething\InvalidRuleException;

abstract class Event extends RuleItem {
	
	
	final public function match(Signal $signal){
		if($this->isValid()){
			
			// reset error
			$this->setError(false);
			
			try {
				return boolval($this->call($signal));
			}
			catch(InvalidRuleException $e){
				// this exception is fired when an action is no more valid.
				$this->setInvalid();
				$this->setError($e->getMessage());
			}
			catch(\Exception $e){
				$this->setError($e->getMessage());
			}
			
		}
		
		return false;
	}
	
	protected function call(Signal $signal) {
		return true;
	}
	
	
}


	

File: ething/rule/Event.py
# coding: utf-8

from . import Signal
from .RuleItem import RuleItem
from . import InvalidRuleException
from ething.meta import MetaEvent
from future.utils import with_metaclass


class Event(with_metaclass(MetaEvent,RuleItem)):
    
    
    def match(self, signal):
        if self.valid:
            
            # reset error
            self.error = False
            
            try:
                return bool(self.call(signal))
            
            except InvalidRuleException as e:
                # this exception is fired when this event is no more valid.
                self.valid = False
                self.error = e
                self.ething.log.exception("Fatal error in event %s" % self)
            
            except Exception as e:
                self.error = e
                self.ething.log.exception("Error in event %s" % self)
        
        return False
    
    
    def call(self, signal):
        return True
    
    @classmethod
    def emit(cls, attribute = {}):
        return Signal(cls.__name__, attribute)
    
